### DISCUSSION TITLE
Cookie not sent with httpSubscriptionLink in tRPC

### DISCUSSION THREAD

**Discussion by masani-dev** [trpc/trpc#6022] [Q&A] [answered]
**Description:**

Hello, I'm implementing Server-Sent Events (SSE) using the `httpSubscriptionLink` as per the [tRPC documentation](https://trpc.io/docs/client/links/httpSubscriptionLink). I need to send cookies for authentication purposes, but it seems they are not being sent.

The documentation mentions:
> If you're doing a web application, you can ignore this section as the cookies are sent as part of the request.

However, I am not seeing any cookies being sent, and I can't find any configuration to ensure cookies are included.

**Setup:**
- **Frontend**: Next.js with the vanilla tRPC client
- **Backend**: Express using the tRPC Express adapter

**Issue:**
- Cookies are not being sent with `httpSubscriptionLink`, and there doesn't seem to be a way to configure cookie sending.
  
**Questions:**
1. How can I configure or ensure that cookies are sent with `httpSubscriptionLink`?
2. Is there any known workaround or additional configuration for sending cookies with SSE?

Any help would be greatly appreciated!

**Comment by KATT** [comment, marked as answer]
Hey! I'm not clear if you want the server to receive auth request or to send headers, so below is both.

## Receiving headers from the client

### Web apps

If you're doing a web application, headers should be sent as part of the request - I just retested this locally (in Chrome) using our [SSE example](https://github.com/trpc/examples-next-sse-chat) and `onAdd:` does indeed have access to the `opts.ctx.session` when the user is logged in.

It might be that you're hosting the backend on a separate domain, in which case you can set `eventSourceOptions: { withCredentials: true }` to `httpSubscriptionLink`.

**Example:**

```ts
import { httpSubscriptionLink } from '@trpc/client';

const link = httpSubscriptionLink({
  url: 'https://example.com/api/trpc',
  eventSourceOptions: {
    withCredentials: true,
  },
});
```

### React Native and similar

- Pass custom headers: https://trpc.io/docs/client/links/httpSubscriptionLink#authorization-by-polyfilling-eventsource
- Send in [`connectionParams`](https://trpc.io/docs/client/links/httpSubscriptionLink#connectionParams)

## Sending headers from the server in a subscription

So, you can't send a cookie in the `.subscription(() => ()`-callback because it's "too late" to send headers then as the headers have already been sent, but you should be able to do it in a middleware.

What you could potentially do is to

```ts
export const appRouter = router({
    onPost: publicProcedure
        .use(opts => {
            // add the cookie to the context
            opts.ctx.res.setHeader('Set-Cookie', 'foo=bar');

            return opts.next()
        })
        .subscription(async function* (opts) {
            // listen for new events
            for await (const [data] of on(ee, 'add')) {
                const post = data as Post;
                yield post;
            }
        );
})
```

---

Please let us know more about your case so we can update the docs (feel free to open a PR for it too :pray:).

**Reply by masani-dev** [reply]
Hi KATT,

Thank you so much for your detailed response! The specific case you mentioned about hosting the backend on a separate domain and using eventSourceOptions: { withCredentials: true } was exactly what I needed. I also appreciate how you thoroughly explained both the client-side and server-side solutions—it was extremely helpful. Your clear guidance made a big difference, and I sincerely appreciate the time you took to assist me.

Just to clarify, my case is indeed a web app, and the backend is hosted on a separate domain.
